### PR TITLE
Generalize broadcast

### PR DESCRIPTION
**Pull request by wsmoses**

### CODE REVIEW

**Comment by wsmoses** [comment]
Companion PR to https://github.com/EnzymeAD/Enzyme/pull/1952

Untested, but should enable arbitrary broadcast support (if all scalar functions are overloaded)

File: src/Reactant.jl [hunk 1]
```diff
@@ -429,7 +429,7 @@ function append_path(path, i)
     return (path..., i)
 end
 
-@inline function make_tracer(seen::IdDict, prev::RT, path, mode, data) where {RT}
+@inline function make_tracer(seen::IdDict, prev::RT, path, mode, data; toscalar=false, tobatch=nothing) where {RT}
```
Comment (mofeing) [inline src/Reactant.jl:432]: What do these kwargs do?

Comment (wsmoses) [reply src/Reactant.jl:432]: whether the tracer we create should automatically change all arrays to scalars, or scalars to arrays

File: src/overloads.jl [hunk 2]
```diff
@@ -457,6 +457,95 @@ for (jlop, hloop) in (
     end
 end
 
+
+function elem_apply(f, args::VarArgs{Nargs})
```
Comment (mofeing) [inline src/overloads.jl:461]: Need `where` here:
```suggestion
function elem_apply(f, args::VarArgs{Nargs}) where {Nargs}
```

File: src/utils.jl [hunk 3]
```diff
@@ -13,7 +13,7 @@ function apply(f, args...; kwargs...)
     return f(args...; kwargs...)
 end
 
-function make_mlir_fn(mod, f, args, kwargs, name="main", concretein=true)
+function make_mlir_fn(mod, f, args, kwargs, name="main", concretein=true; toscalar=false)
```
Comment (mofeing) [inline src/utils.jl:16]: Maybe we could turn `name` and `concretein` into kwargs?
```suggestion
function make_mlir_fn(mod, f, args, kwargs; name="main", concretein=true, toscalar=false)
```

File: src/overloads.jl [hunk 4]
```diff
@@ -457,6 +457,95 @@ for (jlop, hloop) in (
...
+    primf = f.val
+
+    mod = MLIR.IR.module()
```
Comment (mofeing) [inline src/overloads.jl:464]:
```suggestion
    mod = MLIR.IR.mmodule()
```

File: src/overloads.jl [hunk 5]
```diff
@@ -457,6 +457,95 @@ for (jlop, hloop) in (
...
+        end
+    end
+
+
+    function act_attr(val)
+        val = @ccall MLIR.API.mlir_c.enzymeActivityAttrGet(
+            MLIR.IR.context()::MLIR.API.MlirContext, val::Int32
+        )::MLIR.API.MlirAttribute
+        return MLIR.IR.Attribute(val)
+    end
```
Comment (mofeing) [inline src/overloads.jl:511]: This should be declared out of `elem_apply`'s definition if it's not capturing any variable.

**Review by mofeing — CHANGES_REQUESTED** [review]
Just some fix requests and a couple of aesthetic suggestions.

Also, we should add a test for this.

**Comment by avik-pal** [comment]
https://github.com/EnzymeAD/Reactant.jl/blob/5134fe0d2bb6f6d7a4f72c1f0e2d996fe744d0f9/deps/clang#L2 need to be generalized?

File: src/overloads.jl [hunk 6]
```diff
@@ -457,6 +457,95 @@ for (jlop, hloop) in (
...
+
+    fnwrap, func2, traced_result, result, seen_args, ret, linear_args, in_tys, linear_results = make_mlir_fn(
+        mod, f, args, (), string(f) * "_broadcast_scalar", false
+    )
+
+    invmap = IdDict()
+    OutShape = nothing
+    for (k, v) in seen_args
+        invmap[v] = k
+        OutShape size(k)
```
Comment (avik-pal) [inline src/overloads.jl:474]:
```suggestion
        OutShape = size(k)
```

**Pull request merged**